posthog/async_migrations/migrations/0004_replicated_schema.py

```diff
@@ -0,0 +1,305 @@
+
+    @property
+    def renamed_table_name(self):
+        return self.name
+
+    @property
+    def backup_table_name(self):
+        return f"{self.name}_backup_0004_replicated_schema4"
+
+    @property
+    def tmp_table_name(self):
+        return f"{self.name}_tmp_0004_replicated_schema4"
+
+
+@dataclass(frozen=True)
+class ShardedTableMigrationData(TableMigrationData):
+    rename_to: str
+    extra_tables: List[str]
+
+    @property
+    def renamed_table_name(self):
+        return self.rename_to
+
+
+class Migration(AsyncMigrationDefinition):
+
+    description = "Replace tables with replicated counterparts"
+
+    depends_on = "0003_fill_person_distinct_id2"
+
+    def is_required(self):
+        return "Distributed" not in self.get_current_engine("events")
+
+    @cached_property
+    def operations(self):
+        assert settings.CLICKHOUSE_REPLICATION, "CLICKHOUSE_REPLICATION env var needs to be set for this migration"
+        self.validate_number_of_nodes_in_cluster()
+
+        TABLE_MIGRATION_OPERATIONS = [
+            operation for table in self.tables_to_migrate() for operation in self.replicated_table_operations(table)
+        ]
+
+        return [
+            AsyncMigrationOperation.simple_op(sql="SYSTEM STOP MERGES", rollback="SYSTEM START MERGES"),
+            AsyncMigrationOperation(
+                fn=lambda _: setattr(config, "COMPUTE_MATERIALIZED_COLUMNS_ENABLED", False),
+                rollback_fn=lambda _: setattr(config, "COMPUTE_MATERIALIZED_COLUMNS_ENABLED", True),
+            ),
+            *TABLE_MIGRATION_OPERATIONS,
+            AsyncMigrationOperation(
+                fn=lambda _: setattr(config, "COMPUTE_MATERIALIZED_COLUMNS_ENABLED", False),
+                rollback_fn=lambda _: setattr(config, "COMPUTE_MATERIALIZED_COLUMNS_ENABLED", True),
+            ),
+            AsyncMigrationOperation.simple_op(sql="SYSTEM START MERGES", rollback="SYSTEM STOP MERGES",),
+        ]
+
+    def replicated_table_operations(self, table: TableMigrationData):
+        yield AsyncMigrationOperation.simple_op(
+            sql=f"""
+            CREATE TABLE {table.tmp_table_name} AS {table.name}
+            ENGINE = {self.get_new_engine(table)}
+            """,
+            rollback=f"DROP TABLE IF EXISTS {table.tmp_table_name}",
+        )
+
+        if table.materialized_view_name is not None:
+            yield AsyncMigrationOperation.simple_op(
+                sql=f"DROP TABLE IF EXISTS {table.materialized_view_name}", rollback=table.create_materialized_view,
+            )
+
+        yield AsyncMigrationOperation(
+            fn=lambda _: self.move_partitions(table.name, table.tmp_table_name),
+            rollback_fn=lambda _: self.move_partitions(table.tmp_table_name, table.name),
+        )
+        yield AsyncMigrationOperation.simple_op(
+            sql=f"RENAME TABLE {table.name} TO {table.backup_table_name}, {table.tmp_table_name} TO {table.renamed_table_name}",
```

Comment:
If we are backing up, should we also verify that the backups are valid?

Reply:
(I have no idea if that is a valid suggestion tbh but thought worth asking, non-blocking)

Reply:
Not sure what you mean by that. I'm just leaving the old table (with the previous schema) around with `backup_table_name`. Do you have suggestions around the naming?

Reply:
I'm asking here if we could add tests to verify backup tables are present

Reply:
Ah I see! Done!